CMake/Findmlpack.cmake

```
@@ -0,0 +1,641 @@
+    if (NOT compile)
+      message(WARNING "OpenBLAS is downloaded but not compiled. Please compile
+      OpenBLAS before compiling mlpack")
+    else()
+      execute_process(COMMAND make NO_SHARED=1 WORKING_DIRECTORY ${CMAKE_BINARY_DIR}/deps/OpenBLAS-${version})
+      file(GLOB OPENBLAS_LIBRARIES "${CMAKE_BINARY_DIR}/deps/OpenBLAS-${version}/libopenblas.a")
+      set(BLAS_openblas_LIBRARY ${OPENBLAS_LIBRARIES})
+      set(LAPACK_openblas_LIBRARY ${OPENBLAS_LIBRARIES})
+      set(BLAS_FOUND ON)
+    endif()
+  endif()
+
+  find_armadillo(${CMAKE_BINARY_DIR})
+  if (NOT ARMADILLO_FOUND)
+    if (NOT CMAKE_CROSSCOMPILING)
+      find_package(BLAS QUIET)
+      find_package(LAPACK QUIET)
+      if (NOT BLAS_FOUND AND NOT LAPACK_FOUND)
+        message(FATAL_ERROR "Can not find BLAS or LAPACK!  These are required for Armadillo.  Please install one of them---or install Armadillo---before installing mlpack.")
+      endif()
+    endif()
+    get_deps(https://files.mlpack.org/armadillo-${ARMADILLO_VERSION}.tar.gz armadillo armadillo-${ARMADILLO_VERSION}.tar.gz)
+    set(ARMADILLO_INCLUDE_DIR ${GENERIC_INCLUDE_DIR})
+    if (NOT CMAKE_CROSSCOMPILING)
+      find_armadillo(${CMAKE_BINARY_DIR})
+    endif()
+    # Include directories for the previous dependencies.
+    set(MLPACK_INCLUDE_DIRS ${MLPACK_INCLUDE_DIRS} ${ARMADILLO_INCLUDE_DIRS})
+    set(MLPACK_LIBRARIES ${MLPACK_LIBRARIES} ${ARMADILLO_LIBRARIES})
+  endif()
+
+  find_ensmallen(${CMAKE_BINARY_DIR})
+  if (NOT ENSMALLEN_FOUND)
+    get_deps(https://www.ensmallen.org/files/ensmallen-latest.tar.gz ensmallen ensmallen-latest.tar.gz)
+    set(ENSMALLEN_INCLUDE_DIR ${GENERIC_INCLUDE_DIR})
+    find_ensmallen(${CMAKE_BINARY_DIR})
+    if (ENSMALLEN_FOUND)
+      set(MLPACK_INCLUDE_DIRS ${MLPACK_INCLUDE_DIRS} "${ENSMALLEN_INCLUDE_DIR}")
+    endif()
+  endif()
+
+  find_cereal(${CMAKE_BINARY_DIR})
+  if (NOT CEREAL_FOUND)
+    get_deps(https://github.com/USCiLab/cereal/archive/refs/tags/v1.3.2.tar.gz cereal cereal-1.3.2.tar.gz)
+    set(CEREAL_INCLUDE_DIR ${GENERIC_INCLUDE_DIR})
+    find_cereal(${CMAKE_BINARY_DIR})
+    if (CEREAL_FOUND)
+      set(MLPACK_INCLUDE_DIRS ${MLPACK_INCLUDE_DIRS} ${CEREAL_INCLUDE_DIR})
+    endif()
+  endif()
+
+  mlpack(${CMAKE_BINARY_DIR})
+  if (NOT MLPACK_FOUND)
+    get_deps(https://www.mlpack.org/files/mlpack-latest.tar.gz mlpack mlpack-latest.tar.gz)
+    set(MLPACK_INCLUDE_DIR ${GENERIC_INCLUDE_DIR})
+    find_mlpack(${CMAKE_BINARY_DIR})
+    if (MLPACK_FOUND)
+      set(MLPACK_INCLUDE_DIRS ${MLPACK_INCLUDE_DIRS} ${MLPACK_INCLUDE_DIR})
+    endif()
+  endif()
+
+  find_openmp() 
+
+endmacro()
+
+# Download and compile OpenBLAS if we are cross compiling mlpack for a specific
+# architecture. The function takes the version of OpenBLAS as variable.
+macro(fetch_mlpack_and_crosscompile)
+  # 1. Pull openblas and crossocompile it
+  search_openblas(0.3.28)
```

```suggestion
  search_openblas(${OPENBLAS_VERSION})
```
(Assuming that you set a variable with the OpenBLAS version up top.)